modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3Service.java

```diff
@@ -395,6 +400,21 @@ static class CustomWebIdentityTokenCredentialsProvider implements AWSCredentials
                     roleSessionName,
                     webIdentityTokenFileSymlink.toString()
                 ).withStsClient(stsClient).build();
+                var watcher = new FileWatcher(webIdentityTokenFileSymlink);
+                watcher.addListener(new FileChangesListener() {
+
+                    @Override
+                    public void onFileChanged(Path file) {
+                        if (file.equals(webIdentityTokenFileSymlink)) {
+                            credentialsProvider.refresh();
+                        }
```

Comment:
1. Can we have a debug log here?
2. A deletion plus re-creation is basically an update as well. Therefore I think we should handle `onFileCreated` as well. We can skip `onFileDeleted` if we do not want to actively destroy the session. Or maybe we could just log a message for it.